utils/providers.py

```diff
@@ -1,388 +1,391 @@
+        # if not, do not collect the provider for this particular test.
+        if test_flags and test_flags != ['']:
+            test_flags = [flag.strip() for flag in test_flags]
+
+            defined_flags = conf.cfme_data.get('test_flags', '').split(',')
+            defined_flags = [flag.strip() for flag in defined_flags]
+
+            excluded_flags = provider.data.get('excluded_test_flags', '').split(',')
+            excluded_flags = [flag.strip() for flag in excluded_flags]
+
+            allowed_flags = set(defined_flags) - set(excluded_flags)
+
+            if set(test_flags) - allowed_flags:
+                logger.info("Filtering Provider %s out because it does not have the right flags, "
+                            "%s does not contain %s",
+                            provider.data['name'], list(allowed_flags),
+                            list(set(test_flags) - allowed_flags))
+                return False
+        return True
+
+    def _filter_by_version(self, provider):
+        try:
+            if "since_version" in provider.data:
+                # Ignore providers that are not supported in this version yet
+                if version.current_version() < provider.data["since_version"]:
+                    return False
+        except Exception:  # No SSH connection
+            return False
+        return True
+
+    def __call__(self, provider):
+        """ When an instance of this class is called, it will apply this filter on a given provider
+
+        Usage:
+            pf = ProviderFilter('cloud_infra', categories=['cloud', 'infra'])
+            providers = list_providers([pf])
+            pf2 = ProviderFilter(types=['openstack', 'ec2'], required_fields=['small_template'])
+            provider_keys = list_provider_keys([pf, pf2])
+            ...or...
+            pf = ProviderFilter(required_tags=['openstack', 'complete'])
+            pf_inverted = ProviderFilter(required_tags=['disabled'], inverted=True)
+            providers = setup_a_provider([pf, pf_inverted])
+
+        Returns:
+            `True` if provider passed all checks and was not filtered out, `False` otherwise.
+            The result is opposite if the 'inverted' attribute is set to `True`.
+        """
+        if self.keys and provider.key not in self.keys or \
+           self.categories is not None and provider.category not in self.categories or \
+           self.types is not None and provider.type not in self.types or \
+           self.required_fields and not self._filter_required_fields(provider,
+                                                                     self.required_fields) or \
+           self.restrict_version and not self._filter_restricted_version(provider) or \
+           self.required_tags and not self._filter_tags(provider, self.required_tags) or \
+           self.rejected_tags and self._filter_tags(provider, self.rejected_tags):
+            return self.inverted
```

Comment:
Yeh, I would actually prefer to split these into assignments and then run like an ```all()``` because if one of these conditionals is wrong, debugging is going to be a nightmare